lib/nrf_inbuilt_key/nrf_inbuilt_key.c

```diff
@@ -0,0 +1,307 @@
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
+	}
+}
+
+static int write_at_cmd_with_cme_enabled(char *cmd, char *buf, size_t buf_len,
+					 enum at_cmd_state *state)
+{
+	int retval;
+	bool cmee_activated = false;
+
+	int cmee_active_result = cmee_active();
+
+	if (cmee_active_result < 0) {
+		return -EFAULT;
+	}
+
+	if (cmee_active_result == 0) {
+		int result = cmee_set(true);
+
+		if (result != 0) {
+			return result;
+		}
+		cmee_activated = true;
+	}
+
+	retval = at_cmd_write(cmd, buf, buf_len, state);
+
+	if (cmee_activated) {
+		int result = cmee_set(false);
+
+		if (result != 0) {
+			return result;
+		}
+	}
+
+	return retval;
+}
+
+int nrf_inbuilt_key_write(nrf_sec_tag_t sec_tag,
+			  enum nrf_key_mgnt_cred_type_t cred_type,
+			  u8_t *p_buffer, u16_t buffer_len)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((p_buffer == NULL) || (buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_WR) + buffer_len +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 3) +
+				(NRF_INBUILT_KEY_QUOTES_CHAR_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	VALIDATE_LENGTH(scratch_buf_len, sizeof(scratch_buf));
+
+	written = snprintf(scratch_buf, scratch_buf_len, "%s,%u,%u,\"",
+			   NRF_INBUILT_KEY_OP_WR, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+
+	VALIDATE_LENGTH(written, scratch_buf_len);
+
+	VALIDATE_LENGTH(written + buffer_len, scratch_buf_len);
+	memcpy(&scratch_buf[written], p_buffer, buffer_len);
+	written += buffer_len;
+
+	VALIDATE_LENGTH(written + 4, sizeof(scratch_buf));
+	memcpy(&scratch_buf[written], "\"\r\n", 4);
+	written += 4;
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, NULL, 0, &state);
+
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		return translate_error(result);
+	}
+
+	return result;
+}
+
+int nrf_inbuilt_key_read(nrf_sec_tag_t sec_tag,
+			 enum nrf_key_mgnt_cred_type_t cred_type,
+			 u8_t *p_buffer, u16_t *p_buffer_len)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((p_buffer == NULL) || (p_buffer_len == NULL) ||
+	    (*p_buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_RD) +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	VALIDATE_LENGTH(scratch_buf_len, sizeof(scratch_buf));
+
+	written = snprintf(scratch_buf, scratch_buf_len, "%s,%u,%u\r\n",
+			   NRF_INBUILT_KEY_OP_RD, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+
+	if (written < 0) {
+		return -ENOBUFS;
+	}
+
+	VALIDATE_LENGTH(written, scratch_buf_len);
+	VALIDATE_LENGTH(written, *p_buffer_len);
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, scratch_buf,
+					       sizeof(scratch_buf), &state);
```

Comment:
The second param should be NULL

Reply:
I'm not sure which you really mean here. The response buffer? We would need the cmd, response and the CME error in this case.

Reply:
In which code path is the response used? I'm not able to see it.

Reply:
If result is success, it is parsed in the second `if` from here. put into the parser to get the content payload.
```
at_parser_params_from_str(scratch_buf, NULL, &cmng_list);
```